### PR TITLE
Add basic unit tests

### PR DESCRIPTION
**Pull request by hershen**
Meant to be reviewed after #6.

### CODE REVIEW

File: hersh/evaluate.test.cpp
Comment (oorijsonar) [inline hersh/evaluate.test.cpp]: These should probably separate sections. Otherwise I'm not sure why we need both.

File: hersh/hersh.cpp
Comment (oorijsonar) [inline hersh/hersh.cpp]: Shouldn't the default prompt live in the `reader`?

File: hersh/hersh.cpp
Comment (oorijsonar) [inline hersh/hersh.cpp]: like

File: hersh/hersh.cpp
Comment (oorijsonar) [inline hersh/hersh.cpp]: Should probably be initialized with a prompt. A `reader` without a prompt is like a chimney sweep without a chimney sweeper.

File: hersh/printer.test.cpp
Comment (oorijsonar) [inline hersh/printer.test.cpp]: Same as above.

File: hersh/hersh.cpp
Comment (oorijsonar) [inline hersh/hersh.cpp]: like

File: hersh/reader.hpp
Comment (oorijsonar) [inline hersh/reader.hpp]: Can you explain why use `explicit` here? I've never seen it before

File: hersh/reader.test.cpp
Comment (oorijsonar) [inline hersh/reader.test.cpp]: I'm not sure this is a behavior worth testing... Is it important to us that the default prompt is an empty string?

**Review by oorijsonar — APPROVED** [review]

File: hersh/evaluate.test.cpp
Comment (hershen) [reply hersh/evaluate.test.cpp]: I like to always test the empty case.

File: hersh/hersh.cpp
Comment (hershen) [reply hersh/hersh.cpp]: Yes, that's better.

File: hersh/hersh.cpp
Comment (hershen) [reply hersh/hersh.cpp]: I set Reader's default prompt to "hersh$ ". So no need to initialize. Or is that what you meant?

Comment (oorijsonar) [reply hersh/hersh.cpp]: Partly. It's better now. But, it doesn't really make sense to set the prompt after initializing a `Reader` so I would get rid of the setter and add a constructor that allows initializing it.

File: hersh/reader.hpp
Comment (hershen) [reply hersh/reader.hpp]: It prevents the constructor from doing implicit conversions. 
See explanation in [cppreference](https://en.cppreference.com/w/cpp/language/explicit) and the examples at the bottom which are illuminating, and also [stackoverflow](https://stackoverflow.com/questions/121162/what-does-the-explicit-keyword-mean). They have good explanations. I've never used it before, but it's intended to prevent unintended and hard to find bugs. We can try it see if it's useful.

You might say that it's not needed for empty constructors, and that's true. But I have a feeling it's a good habit to make every constructor explicit.

File: hersh/hersh.cpp
Comment (hershen) [reply hersh/hersh.cpp]: I haven't read the whole sh specifications, but if it's similar to the default bash prompt behavior, the prompt changes each time you change directory. So I think we'll have to change the prompt a lot.

File: hersh/reader.test.cpp
Comment (hershen) [reply hersh/reader.test.cpp]: I think it's important to test that the class does what you think it does. If you think the default prompt is empty, you should check that. That's how I see it.

Comment (oorijsonar) [reply hersh/reader.test.cpp]: That means that every time we change the default prompt we need to change the test. I think this test causes more problems than it solves.

File: hersh/hersh.cpp
Comment (oorijsonar) [reply hersh/hersh.cpp]: 👍

File: hersh/evaluate.test.cpp
Comment (oorijsonar) [reply hersh/evaluate.test.cpp]: That means you think it's a special case therefore it should have its own section.

Comment (hershen) [reply hersh/evaluate.test.cpp]: Partitioning into SECTIONS causes everything between TEST_CASE and the first SECTION to be run before each SECTION. Like a set-up. I don't think there's a need for set-ups here (yet?).

And why should special cases be in a different section even if there's no set-up issues?

File: hersh/reader.test.cpp
Comment (hershen) [reply hersh/reader.test.cpp]: Yes, you'd have to change the test, because you change the class. I don't see a problem with that. 

In any case, I don't think we'll change the default prompt a lot. Maybe one more time to make it more consistent with bash ( just '$' instead of 'hersh$').

Comment (oorijsonar) [reply hersh/reader.test.cpp]: When is this test useful? When you make a typo in the default prompt? It causes trouble and will never catch any bugs.

File: hersh/evaluate.test.cpp
Comment (oorijsonar) [reply hersh/evaluate.test.cpp]: I'm used to organizing tests in individual test cases and it seems like you're using sections as if they separate test cases. Adding addition assertions (`REQUIRES` in this case) should be done for each test case individually. The test case's name indicates what you're testing.
Therefore having a test case named "check empty case" and one named "check happy path" would be more useful in this case. That way when a test case fails you understand what it's supposed to test immediately by reading its name. I don't know if they should be separate `SECTION`s or separate `TEST_CASE`s but they should be separate something.

Comment (hershen) [reply hersh/evaluate.test.cpp]: I think that makes sense. I was following the example in the [Catch2 tutorial](https://github.com/catchorg/Catch2/blob/devel/docs/tutorial.md#test-cases-and-sections) is not a good example?
Are you saying it's not a good example?

Edit: I think I understand - partition with either TEST_CASE or with SECTION.

Comment (oorijsonar) [reply hersh/evaluate.test.cpp]: 👍

File: hersh/printer.test.cpp
Comment (hershen) [reply hersh/printer.test.cpp]: Done.

File: hersh/evaluate.test.cpp
Comment (hershen) [reply hersh/evaluate.test.cpp]: Done. There's no nested TEST_CASE, so you were right, it needs to be partitioned with SECTIONs.

File: hersh/reader.test.cpp
Comment (hershen) [reply hersh/reader.test.cpp]: Done.

**Comment by hershen** [comment]
All comments addressed. Waiting for re-approval.

**Review by oorijsonar — APPROVED** [review]

**Pull request merged**
